### ISSUE TITLE
How do I style bold in code blocks?

### ISSUE THREAD

**Issue by hamishwillee** [mdn/yari#2198]
https://developer.mozilla.org/en-US/docs/Learn/Server-side/Django/skeleton_website

This bit used to be bold: 
![image](https://user-images.githubusercontent.com/5368500/102559734-b23e7d80-4124-11eb-95c6-c34aa84bf0ce.png)

It is marked up as shown
```html
<pre class="brush: bash notranslate">INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
<strong>    'catalog', </strong>
]</pre>
```

**Comment by schalkneethling** [comment]
@hamishwillee From looking at the example now I believe what is different might be that not _all_ code blocks was syntax highlighted before but now, they are. This also happens server-side before the HTML is sent to the browser(@peterbe correct me if I am wrong here).

I am not sure what the workaround would be at the moment. Perhaps we do not have one right this moment. I guess instead of processing all `pre` elements via Prismjs, we could change it so it only processes `pre` elements with a specific class. With that said, this one does contain `brush: bash notranslate` so, it seems like it is supposed to be syntax highlighted.

Could you previous have a code block be syntax highlighted but still get some of the content to be marked as `bold`?

**Comment by peterbe** [comment]
Short answer is; you can't at the moment. 

The way it works is that we grab each `<pre>` tag, and get the **text** out of it. Kinda like you'd do with your mouse cursor were you select it all and copy and paste into notepad. All formatting is gone. Once we have the pure text, we send it into the Node version of PrismJS and outcomes a blob of `<span class="token punctuation">;</span><span class="token ...` etc. We then put that back into the `<pre>` as a trusted HTML blob. That's why your HTML tags are lost. 

Truthfully, the Node implementation of Prism is worse than that for client-side (browser DOM) rendering. It's annoying. But a not-so-subtle advantage is that HTML usually renders much faster than JavaScript. And it compresses well too. Now, all you need to render the code snippets is HTML + CSS. But in the process we lost some of fancy features that the client-side Prism had such as specifying specific lines to have a bold background, for standing out more. We also lost the ability to toggle line-numbers on and off :(

And another interesting war-story is that there's a LOT of HTML like this in the content: https://github.com/mdn/content/blob/b20727e08889a26686299b3cca9517b0a516770d/files/en-us/learn/css/building_blocks/the_box_model/index.html#L155-L160
I suspect someone copied the rendered Prism HTML (perhaps from the DOM inspector or something). There's even a flaw for spotting these and cleaning them up. 

However, we could get smarter and for time-strategy reasons, we have to make a choice:

1. Come up with a white-list solution of allowing and respecting some HTML tags within the `<pre>` tags and preserve them after Node Prism has done its thing. 
2. Figure out how to get the `highlight-line` Prism plugin to work in Node. Can we pay the maintainers of Prism or is possible with the right amount of elbow grease??
3. A more profound solution where you annotate a specific keyword and using the HTML you can technically get to it with something like `span.token.keyword:nth-child(7){ font-weight: bold }`. 

Does Markdown support the ability to highlight specific lines?
I think we can hack around the Node Prism highlight-line plugin shortcomings since we can loop over the lines in the rendered HTML, but that's not something we have just yet.

**Comment by mattiapontonio** [comment]
## This should be considered

[<code>: The `Inline` Code element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/code)

_To represent multiple lines of code, wrap the <code> element within a <pre> element. The <code> element by itself only represents a single phrase of code or line of code._

```html
<pre>
<code class="marked-as-important-by-the-editor bad-line-red" />code;</code>
</pre>
```

**Comment by hamishwillee** [comment]
@peterbe Thanks for the detailed explanation

> Does Markdown support the ability to highlight specific lines?

Generic github markdown does not, but various other extensions and libraries support it. I find it a useful feature. In systems where I don't have the feature, I tend to do my highlighting by displaying multiple blocks or using comments. In this case multiple lines look silly. E.g. 
```
INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',
```
```
'catalog',
```
```
]
```

From my perspective I just need to know if you think it is something I should go through and fix in some other way, or which might be supported in the next month or two so I should wait!

**Comment by mattiapontonio** [comment]
The bad thing about comments in code snippets is that you have to delete them when you copy them into yours.

**Comment by hamishwillee** [comment]
@mattiapontonio Not really - the point is to use code comments for the language, so that the code still compiles if copied.

**Comment by mattiapontonio** [comment]
@hamishwillee i was saying that sometimes I do

```JavaScript
const a = 1;
const b = 2; // <-- look at this line
```

but i would prefer

```JavaScript
const a = 1;
**const b = 2;**
```

**Comment by mattiapontonio** [comment]
FTR
https://meta.stackexchange.com/q/183610/580790

**Comment by hamishwillee** [comment]
@mattiapontonio Thanks for clarification. Yes, that's kind of what I was saying - I can mark up in comments but it isn't my preference.

@peterbe When you're back from holiday, as above "I just need to know if this will be fixed or not, and if so, in a timely manner". That will tell me if I need to modify docs as I work on them, or leave them to be fixed automatically in future.

**Comment by mattiapontonio** [comment]
<pre>
const a = 1;
<strong>const b = 2;</strong>
</pre>

I discovered that You can use HTML in markdown here in GitHub.

**Comment by mattiapontonio** [comment]
There is a basic problem though: that when you insert bash into a pre you make a kind of agreement to insert bash and not HTML. I think the parser from the pre class expects a certain system from which he then adds the colored classes. If you mix languages I don't know if he understands.

The editor should use a normal HTML pre and add the colored classes for a perfect, highlighted and colored result.

At the moment, an easy solution does not occur to me except manually creating these HTML which, however, as I understand it, can not be embedded in an iframe result.

**Comment by mattiapontonio** [comment]
@all I'm curious about this one.
@hamishwillee do You know other projects that has this solution of bolding lines of code blocks?

**Comment by Ryuno-Ki** [comment]
Since I'm dealing with reveal.js right now, here's how they do it:

https://github.com/hakimel/reveal.js/blob/ffadcc850266099c03709dd52865f93040fa602c/plugin/markdown/plugin.js#L434-L442

That would lead to data attributes triggering the generation of fragments:

https://github.com/hakimel/reveal.js/blob/ffadcc850266099c03709dd52865f93040fa602c/plugin/highlight/plugin.js#L98-L137

The generated code looks like this:

```html
<pre>
  <code data-line-numbers="3,8-10" data-trim="" data-noescape="" class="hljs xml has-highlights">
    <table class="hljs-ln">
      <tbody>
        <tr>
          <td class="hljs-ln-numbers">
            <div class="hljs-ln-line hljs-ln-n" data-line-number="1"></div>
          </td>
          <td class="hljs-ln-code">
            <div class="hljs-ln-line">
              <span class="hljs-tag">
                &lt;<span class="hljs-name">table</span>&gt;
              </span>
            </div>
          </td>
        </tr>
        <tr>
          <td class="hljs-ln-numbers">
            <div class="hljs-ln-line hljs-ln-n" data-line-number="2"></div>
          </td>
          <td class="hljs-ln-code">
            <div class="hljs-ln-line">
                <span class="hljs-tag">
              &lt;<span class="hljs-name">tr</span>&gt;
              </span>
            </div>
          </td>
        </tr>
        <tr class="highlight-line">
          <td class="hljs-ln-numbers">
            <div class="hljs-ln-line hljs-ln-n" data-line-number="3"></div>
          </td>
          <td class="hljs-ln-code">
            <div class="hljs-ln-line">
                  <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                Apples
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="4"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                    <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                $1
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="5"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                    <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                7
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="6"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                  <span class="hljs-tag">
                  &lt;/<span class="hljs-name">tr</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="7"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                  <span class="hljs-tag">
                  &lt;<span class="hljs-name">tr</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr class="highlight-line">
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="8"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                    <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                Oranges
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr class="highlight-line">
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="9"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                    <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                $2
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr class="highlight-line">
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="10"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                    <span class="hljs-tag">
                  &lt;<span class="hljs-name">td</span>&gt;
                </span>
                18
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">td</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="11"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                  <span class="hljs-tag">
                  &lt;/<span class="hljs-name">tr</span>&gt;
                </span>
              </div>
            </td>
          </tr>
          <tr>
            <td class="hljs-ln-numbers">
              <div class="hljs-ln-line hljs-ln-n" data-line-number="12"></div>
            </td>
            <td class="hljs-ln-code">
              <div class="hljs-ln-line">
                <span class="hljs-tag">
                  &lt;/<span class="hljs-name">table</span>&gt;
                </span>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </code>
  </pre>
```

Taken from https://revealjs.com/code/#line-numbers-%26-highlights

**Comment by mattiapontonio** [comment]
The `<table>` Is good for the line numbers that are not currently supported but handy.

**Comment by Ryuno-Ki** [comment]
I'm worried about screenreader UX when using layout tables :-S

**Comment by mattiapontonio** [comment]
> I'm worried about screenreader UX when using layout tables :-S

Yes.
If You keep font properties consistent You can think on some layout like

🧂
```HTML
<div>
1
2
</div>
<pre>
const a;
const b;
</pre>
```

**Comment by Ryuno-Ki** [comment]
But that would read in a screenreader as „1 2 const a const b” I guess.
Instead, the line numbers could be generated content (via CSS).
Perhaps in conjunction with [CSS counters](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Lists_and_Counters/Using_CSS_counters) if we could identify each line.

**Comment by Ryuno-Ki** [comment]
(That still does not clarify the original question about highlighting certain lines. Here, I could imagine a JavaScript injected `<style>` which uses `nth-child` selector)

**Comment by mattiapontonio** [comment]
Yes.
`nth-child`
not
_nth-line_

**Comment by Ryuno-Ki** [comment]
Okay, let me rephrase.

Given

```html
<pre>
​<code>INSTALLED_APPS = [</code>
<code>'django.contrib.admin',</code>
<code>'django.contrib.auth',</code>
<code>'django.contrib.contenttypes',</code>
<code>'django.contrib.sessions',</code>
<code>'django.contrib.messages',</code>
<code>'django.contrib.staticfiles',</code>
<code>'catalog'</code>
<code>​]​</code>
</pre>
```
if I add a data-attribute (which some Markdown parser allow) to the `<pre>` to highlight the line ninth line (`]`), I can target it via `pre > code:nth-child(9) { font-weight: bold; }`.

This style would have to be added via JavaScript

1. Check `<pre>` for a well-named data-attribute.
2. If present, create a new `<style>` and inject the above rule (`9` is be replaced with proper values - comma separated lines or even ranges, which would be translated into comma-separated selectors)
3. Add the style to the rendered Markdown (perhaps scoped to the respective `<pre>`).

**Comment by mattiapontonio** [comment]
It's a possibile solution but what Is given here Is
https://github.com/mdn/yari/issues/2198#issue-770475607

**Comment by hamishwillee** [comment]
> do You know other projects that has this solution of bolding lines of code blocks?

No :-). 
FYI From a user story point of view all I care about is "I should be able to highlight specific lines of code" - bold is just what the old wiki allowed. 
I won't be adding code blocks (if I can avoid it) until yari supports markdown. No one should have to.

**Comment by mattiapontonio** [comment]
> bold is just what the old wiki allowed

How, a macro? I wasn't here.
Yes the story is good.
I would try https://github.com/mdn/yari/issues/2198#issuecomment-765490957 in a PR on content and see what happens but without local build i don't know what Will be produced. 🤔

**Comment by hamishwillee** [comment]
> How, a macro? I wasn't here.

No, the wiki had a wysiwyg editor which would apply whatever formatting was allowed. The yari migration to a github backend lost us that, which is why enabling markdown in the next iteration is so important. On the plus side I love the tools github offers for change control and diffing.

**Comment by mattiapontonio** [comment]
https://github.com/mdn/content/blob/21e72af0578091c8015ff46ec4426697ebf43210/files/en-us/learn/server-side/django/skeleton_website/index.html#L160

**Comment by mattiapontonio** [comment]
@hamishwillee I'm sorry i don't know now. What o would try Is ti _do it in HTML_ and discuss in the PR via _normal flow_.

**Comment by hamishwillee** [comment]
@mattiapontonio What I am _going_ to do is ignore this until @peterbe responds to https://github.com/mdn/yari/issues/2198#issuecomment-751558347 . That will tell me whether I need to update the docs to compensate for lack of bold now.

**Comment by schalkneethling** [comment]
> "I should be able to highlight specific lines of code"

I wonder if this is possible with Prism on the server-side. I know there is a Prism plugin[1] for indicating that a specific line should be highlighted. But, I know there are(might be?) some limitations on what is possible on the server-side.

[1] https://prismjs.com/plugins/line-highlight/

**Comment by mattiapontonio** [comment]
Antoher bad solution:

1. Realise a normal pre and let Prism work.
2. Go to PROD and hard-copy the HTML
3. Ri-edit the page, removing the Prism classes and Just wrap the lines with strong.

Total time: ⌚ some Days

**Comment by mattiapontonio** [comment]
> the wiki had a wysiwyg editor which would apply whatever formatting was allowed

Let me doubt this without certainty that it would have been possible without errors.

**Comment by hamishwillee** [comment]
@schalkneethling Where is the best place to add line highlighting this as a requirement on our markdown solution? (i.e. I have given up on this being fixed in HTML any time soon, but I want it to be an option  in markdown.

**Comment by schalkneethling** [comment]
> @schalkneethling Where is the best place to add line highlighting this as a requirement on our markdown solution? (i.e. I have given up on this being fixed in HTML any time soon, but I want it to be an option in markdown.

I would open an issue on Yari as the relevant plugin will need to be added to Prismjs. 
https://github.com/mdn/yari/issues/2198#issuecomment-766568687

**Comment by Ryuno-Ki** [comment]
That being said, perhaps leave an HTML comment so we can look up which line should be highlighted (in case the current HTML gets cleaned).

**Comment by mattiapontonio** [comment]
- https://prismjs.com/plugins/line-highlight/#play.8-10
-  https://github.com/mdn/content/blob/d33756432e0b336d33c1f47bb3d0f65e0ce56659/files/en-us/web/css/using_css_custom_properties/index.html#L53

This highlight _substring of the class_ works?
Seems not 🙁 in the rendered HTML. 

---

In any case it happened in my life that I saw lines highlighted in MDN, I don't think I dreamed it.

**Comment by hamishwillee** [comment]
@schalkneethling Thanks. I've created issue for the markdown discussion in https://github.com/mdn/yari/issues/2757

@mattiapontonio The wiki used to allow some forms of markup - e.g. you could use `<strong>`. That's gone now. While I appreciate you trying to find alternatives that somehow work with the current HTML that is not what we need. What we need is an "official/supported mechanism that is simple to use". Anything else is a hack that may well be broken in future.

**Comment by mattiapontonio** [comment]
I still doubt that the old editor would allow bold in the preamps without problems but let's forget about the quirks.
So here the fact lies in the addiction, you have to think about whether to trust the addiction a lot with respect to ourselves and also perhaps you should think about computing those preamps on the Front End (just because the addiction seems to offer functionality not in Node).

**Comment by mattiapontonio** [comment]
Let's forget the past, let's think about the future instead.
2 things (to be explored):
- Use a plugin
- Move the computation to the client

**Comment by schalkneethling** [comment]
@wbamberg @ddbeck Here also, is this something we want to pursue or, have we decided to drop support?

**Comment by wbamberg** [comment]
I think we have decided not to bring this with us into Markdown.

**Comment by hamishwillee** [comment]
@wbamberg Yes, that was agreed. The discussion "concluded" that if relevant lines cannot be described in text without highlighting, then in most cases the root cause is probably that the blocks are too long and should be broken up.

I disagreed on the basis that there are cases where the ability to highlight a particular line of code saves you lots of words and makes things easier for readers. Whether there are enough to justify the exception to github flavoured markdown is arguable. Anyhow, yes, this can be closed.
